### PR TITLE
Fix timezone issues with datetime type

### LINKED ISSUE #105: Timezone issue: after before does not work property
I have created a jsonb column as follows

```
jsonb_accessor :monitoring,
    heartbeat_at: :datetime
```

My rails default timezone is TZInfo::Timezone.get('Asia/Kolkata'), so when I save 

```
monitoring->heartbeat_at = Time.now // This is +5:30 timezone
```

Upon searching via `monitoring_where(heartbeat_at: {after: 2.minute.ago.utc}).count` it wont find the record.

Instead when I save it via 

```
monitoring->heartbeat_at = Time.now.utc 
```

the search query works. Basically the sql query has `(table.monitoring ->> 'heartbeat_at')::timestamp` which doesnot have timezone component in to it.

### PR DESCRIPTION
**Pull request by caiohsramos**
Fixes #105 

Bug description at linked issue

### CODE REVIEW

**Comment by haffla** [comment]
Hello @cainwatson thank you so much for your contribution. I can see that your PR fixes the issue and I am very inclined to just hit "Merge" and be done with it. The thing is that I was hoping for a PR that fixes the issue from the ground up.

Rails by default saves timestamps in UTC. It also converts all datetime objects in `where` to UTC. This of course only unless `config.active_record.time_zone_aware_attributes = false` because by default all `:datetime` and `:time` attributes are time zone aware. So while your PR fixes the problem for most people it doesn't fix it for people who have disabled this feature.

It is now around 17:00 here in Berlin (UTC +2). So UTC time now is 15:00.

```
# time_zone_aware_attributes ENABLED (default)
Book.where('created_at < ', DateTime.current)
#=> SELECT "books".* FROM "books" WHERE (created_at < '2021-05-25 15:01:52.021088')

# time_zone_aware_attributes enabled
Book.where('created_at < ', DateTime.current)
#=> SELECT "books".* FROM "books" WHERE (created_at < '2021-05-25 17:01:52.021088')
```

So I think the real underlying issue here is that this gem does not respect this Rails setting and simply saves datetime objects as is. We need to make this Gem read this Rails setting and save the value accordingly.

**Comment by caiohsramos** [comment]
Thanks for the response @haffla, and for explaining this behaviour. I will try to work later on this today to make sure the gem saves jsonb datetime fields in UTC. `{"sold_at"=>"2021-05-24T20:32:04.890-03:00"}` should be `{"sold_at"=>"2021-05-24T23:32:04.890"}`, right?

PS: I think you mentioned the wrong username in your comment, mine is @caiohsramos :smile:

**Comment by haffla** [comment]
Ah sorry yes @caiohsramos. Sorry about that.

You are absolutely right. It should be `{"sold_at"=>"2021-05-24T23:32:04.890"}` in your case.

**Comment by caiohsramos** [comment]
Hello @haffla. So, I updated my code. 

Doing some research I concluded that:
- Rails saves time-like objects in the database based on the config `config.active_record.default_timezone`. By default is UTC, but may be in the local timezone
- `time_zone_aware_attributes` is probably not the problem here. It works just fine, converting (or not) whatever the dabase returned

Does it makes sense to you?

File: lib/jsonb_accessor/macro.rb [hunk 1]
```diff
@@ -65,7 +65,15 @@ def jsonb_accessor(jsonb_attribute, field_types)
           names_and_store_keys.each do |name, store_key|
             define_method("#{name}=") do |value|
               super(value)
-              new_values = (public_send(jsonb_attribute) || {}).merge(store_key => public_send(name))
+
+              attribute_value = public_send(name)
+              # Rails always saves time based on `default_timezone`. Since #as_json considers timezone, manual conversion is needed
+              if attribute_value.acts_like?(:time)
+                default_timezone = ActiveRecord::Base.default_timezone
+                attribute_value = (default_timezone == :utc ? attribute_value.utc : attribute_value.in_time_zone).strftime("%FT%R:%S.%LZ")
```
Comment (haffla) [inline lib/jsonb_accessor/macro.rb:73]: I think `.strftime("%FT%R:%S.%LZ")` is equivalent to `.iso8601`.

File: lib/jsonb_accessor/macro.rb [hunk 2]
```diff
@@ -65,7 +65,15 @@ def jsonb_accessor(jsonb_attribute, field_types)
           names_and_store_keys.each do |name, store_key|
             define_method("#{name}=") do |value|
               super(value)
-              new_values = (public_send(jsonb_attribute) || {}).merge(store_key => public_send(name))
+
+              attribute_value = public_send(name)
+              # Rails always saves time based on `default_timezone`. Since #as_json considers timezone, manual conversion is needed
+              if attribute_value.acts_like?(:time)
```
Comment (haffla) [inline lib/jsonb_accessor/macro.rb:71]: What if `value` / `attribute_value` is nil? Don't we have a test for this?

**Review by haffla — CHANGES_REQUESTED** [review]

**Comment by haffla** [comment]
mmh no, Rails never saves something like `2021-05-27T12:29:11+02:00` to the database.

For example:

```ruby
book.touch
# or
book.update(updated_at: DateTime.yesterday)
```

Check the generated SQL query and compare it to your local time and repeat with different settings for `config.active_record.default_timezone`. In any case what ends up in the database will be something like `2021-05-27 12:26:05`.

I am afraid this is a very complicated issue to solve.

File: lib/jsonb_accessor/macro.rb (hunk 1 above)
Comment (caiohsramos) [reply lib/jsonb_accessor/macro.rb:73]: Not quite. `#iso8601` has timezone information, this formatter "discards" TZ
```
irb(main):031:0> Time.zone.now.strftime("%FT%R:%S.%LZ")
=> "2021-05-27T09:14:23.876Z"
irb(main):032:0> Time.zone.now.iso8601
=> "2021-05-27T09:14:35-03:00"
```

**Comment by caiohsramos** [comment]
> Rails never saves something like 2021-05-27T12:29:11+02:00

Agreed, it would be:
- When `active_record.default_timezone` is UTC, saves something like `2021-05-27 10:29:11`
- When `active_record.default_timezone` is local, saves something like `2021-05-27 12:29:11`

To keep this behaviour I purposed using a formatter that does not consider TZ as `strftime("%FT%R:%S.%LZ")`

**Comment by haffla** [comment]
But `strftime("%FT%R:%S.%LZ")` always leaves the `Z` at the end which stands for `UTC`.

It's now 15:07 in Berlin. When I do `DateTime.current.strftime("%FT%R:%S.%LZ")` I get `"2021-05-27T15:07:51.044Z"` which says 15:07 in UTC which is not true. I think what you want is `.to_s(:db)`.

**Comment by caiohsramos** [comment]
Saving with `.strftime("%F %R:%S")` would loose milliseconds information, right? 😕 I'll try to work on more tests to this feature, changing `active_record.default_timezone` and `time_zone_aware_attributes`. Then, we can focus on the implementation

**Comment by haffla** [comment]
Yes correct. `.strftime("%F %R:%S:%L")` preserves milliseconds. Though `.to_s(:db)` [doesn't have](https://apidock.com/rails/ActiveSupport/TimeWithZone/to_s) milliseconds either.

P.S.: Don't use `.to_s(:db)`. It converts to UTC.

**Comment by caiohsramos** [comment]
Hey @haffla. I did a commit and some tests to remember what was left to do here, the results were:
* Datetime fields are saved as strings with the format `%FT%R:%S.%L` (no timezone or UTC information)
* The implementation respects `active_record.default_timezone`, both on querying and on persisting
* The implementation does not interfere with the `active_record.time_zone_aware_attributes` configuration. It only matters when the attributes are instantiated, and it is working

Some tests changing the `default_timezone` config would be great

**Review by haffla — CHANGES_REQUESTED** [review]
Wow! This is looking great. Thank you for your contribution. I am quite confident that this fixes the issue. Just a couple of small things to address.

And two more things:

1. I think this needs more test cases. Can you write some? Probably they should go here: spec/jsonb_accessor_spec.rb in it's own `describe` block. This is not about the query builder. It's about the attribute. We need to at least cover the two configs for `default_timezone`: `:utc` and `local`. Make sure the correct strings are saved in the JSON.

File: spec/jsonb_accessor_spec.rb [hunk 3]
```diff
@@ -562,6 +562,31 @@ def build_class(jsonb_accessor_config, &block)
...
+      Time.new(2022, 1, 1, 12, 5, 0, "-03:00")
+    end
+    it "saves in UTC" do
+      instance.foo = time_with_zone
+      expect(instance.options).to eq({ "foo" => time_with_zone.utc.strftime("%F %R:%S.%L") })
+    end
+
+    context "when default_timezone is local" do
+      around(:each) do |example|
+        ActiveRecord::Base.default_timezone = :local
```
Comment (haffla) [inline spec/jsonb_accessor_spec.rb:579]: Same problem here :) This will fail in Rails 7.1.

File: spec/jsonb_accessor_spec.rb [hunk 4]
```diff
@@ -562,6 +562,31 @@ def build_class(jsonb_accessor_config, &block)
...
+  context "datetime field" do
+    let(:klass) do
+      build_class(foo: :datetime)
+    end
+    let(:time_with_zone) do
+      Time.new(2022, 1, 1, 12, 5, 0, "-03:00")
+    end
+    it "saves in UTC" do
+      instance.foo = time_with_zone
+      expect(instance.options).to eq({ "foo" => time_with_zone.utc.strftime("%F %R:%S.%L") })
```
Comment (haffla) [inline spec/jsonb_accessor_spec.rb:574]:
```suggestion
      expect(instance.options).to eq({ "foo" => "2022-01-01 15:05:00.000" })
```

File: spec/jsonb_accessor_spec.rb [hunk 5]
```diff
@@ -562,6 +562,31 @@ def build_class(jsonb_accessor_config, &block)
...
+
+    context "when default_timezone is local" do
+      around(:each) do |example|
+        ActiveRecord::Base.default_timezone = :local
+        example.run
+        ActiveRecord::Base.default_timezone = :utc
+      end
+      it "saves in local time" do
+        instance.foo = time_with_zone
+        expect(instance.options).to eq({ "foo" => time_with_zone.strftime("%F %R:%S.%L") })
```
Comment (haffla) [inline spec/jsonb_accessor_spec.rb:585]:
```suggestion
        expect(instance.options).to eq({ "foo" => "2022-01-01 12:05:00.000" })
```

**Review by haffla — CHANGES_REQUESTED** [review]

File: lib/jsonb_accessor/macro.rb [hunk 6]
```diff
@@ -65,7 +66,14 @@ def jsonb_accessor(jsonb_attribute, field_types)
...
+                attribute_value = (active_record_default_timezone == :utc ? attribute_value.utc : attribute_value.in_time_zone).strftime("%F %R:%S.%L")
```
Comment (haffla) [inline lib/jsonb_accessor/macro.rb:73]: Maybe to be in line with the rest of the code something like `JsonbAccessor::Helpers.active_record_default_timezone` would be better.

**Review by haffla — CHANGES_REQUESTED** [review]

**Comment by haffla** [comment]
@caiohsramos sorry for being so picky, last change requests now before merging this. Thank you so much!

**Comment by caiohsramos** [comment]
@haffla thanks for the review :) Fixed the points you listed

**Review by haffla — APPROVED** [review]

**Pull request merged**